### PR TITLE
[stdlib] `KeyPathIterable` fixes.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Change `Dictionary.allKeyPaths` to return `WritableKeyPath` instances.
  - Previously, `\Dictionary[_: Index].value` was used. Those key paths are not
    writable.
- Assert that `{Array,Array.DifferentiableView,Dictionary}.allKeyPaths` return
  `WritableKeyPath` instances.
- Re-add `test/stdlib/KeyPathIterable.swift`.
  - This test existed at [`test/TensorFlowRuntime/key_path_iterable.swift`](https://github.com/apple/swift/blob/tensorflow-0.2/test/TensorFlowRuntime/key_path_iterable.swift)
    but was deleted.
- Add negative test for TF-123: `@differentiable` function opaque reabstraction.
- Minor gardening included.

---

Examples:

```swift
// Context:
//
// `Dictionary` has two subscripting interfaces:
//
// 1. Subscripting with a key, yielding an optional value:
//   v = d[k]!
// 2. Subscripting with an index, yielding a key-value pair:
//   (k, v) = d[i]
//
// Only (1) can be used to form a `WritableKeyPath`. (2) cannot.

print(\Dictionary<String, Int>["key"]!) // key
// Swift.WritableKeyPath<Swift.Dictionary<Swift.String, Swift.Int>, Swift.Int>

let dummyIndex: Dictionary<String, Int>.Index = [:].startIndex
print(\Dictionary<String, Int>[dummyIndex].value) // index
// Swift.KeyPath<Swift.Dictionary<Swift.String, Swift.Int>, Swift.Int>
```
```swift
// Test:
var dict = ["a": 1, "b": 2, "c": 3]
for kp in dict.allWritableKeyPaths(to: Int.self) {
  dict[keyPath: kp] += 1
}
print(dict)
// Before: ["a": 1, "b": 2, "c": 3]
// After: ["a": 2, "b": 3, "c": 4]

```

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: stdlib/public/core/KeyPathIterable.swift [hunk 1]
```diff
@@ -97,30 +97,56 @@ public extension KeyPathIterable {
 // Collection conformances
 //===----------------------------------------------------------------------===//
 
-extension Array : KeyPathIterable {
+/// Returns `true` if all of the given key paths are instances of
+/// `WritableKeyPath<Root, Value>`.
+private func areAllKeyPathsWritable<Root, Value>(
```
Comment (rxwei) [inline stdlib/public/core/KeyPathIterable.swift]:
```suggestion
private func areWritable<Root, Value>(
```

File: stdlib/public/core/KeyPathIterable.swift [hunk 2]
```diff
@@ -97,30 +97,56 @@ public extension KeyPathIterable {
...
+  _ keyPaths: [PartialKeyPath<Root>], toValueType valueType: Value.Type
```
Comment (rxwei) [inline stdlib/public/core/KeyPathIterable.swift]:
```suggestion
  _ keyPaths: [PartialKeyPath<Root>], valueType: Value.Type
```

**Review by rxwei — APPROVED** [review]

File: stdlib/public/core/KeyPathIterable.swift (hunk 2 above)
Comment (dan-zheng) [reply stdlib/public/core/KeyPathIterable.swift]: Done, thanks.

File: stdlib/public/core/KeyPathIterable.swift (hunk 1 above)
Comment (dan-zheng) [reply stdlib/public/core/KeyPathIterable.swift]: Done.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**